### PR TITLE
GH-35247: [C++] Add Arrow Substrait support for stddev/variance

### LINKED ISSUE #35247: [C++] Add Arrow Substrait support for stddev/variance
### Describe the enhancement requested

Recently, ibis-substrait added [support for stddev/variance](https://github.com/ibis-project/ibis-substrait/pull/583/). This issue will add support for the same in Arrow Substrait.

### Component(s)

C++

### PR DESCRIPTION
**Pull request by rtpsw**
See #35247
* Closes: #35247

### CODE REVIEW

**Review by icexelloss — APPROVED** [review]
LGTM

**Review by ianmcook — CHANGES_REQUESTED** [review]
This should pass `VarianceOptions` with `ddof` set to `1` when the `distribution` option is set to `POPULATION`

**Comment by rtpsw** [comment]
> This should pass `VarianceOptions` with `ddof` set to `1` when the `distribution` option is set to `POPULATION`

Good catch. I'll work on adding this.

File: cpp/src/arrow/engine/substrait/extension_set.cc [hunk 1]
```diff
@@ -944,6 +944,26 @@ ExtensionIdRegistry::SubstraitAggregateToArrow DecodeBasicAggregate(
                                call.id().name, " to have at least one argument");
       }
       case 1: {
+        std::shared_ptr<compute::FunctionOptions> options = nullptr;
+        if (arrow_function_name == "stddev" || arrow_function_name == "variance") {
+          auto maybe_dist = call.GetOption("distribution");
```
Comment (icexelloss) [inline cpp/src/arrow/engine/substrait/extension_set.cc:949]: Can you add a link to the substrait spec here:
https://github.com/substrait-io/substrait/blob/73228b4112d79eb1011af0ebb41753ce23ca180c/extensions/functions_arithmetic.yaml#L1240

**Comment by icexelloss** [comment]
Minor comments otherwise looks good

**Comment by icexelloss** [comment]
@ianmcook @westonpace Do you want to take a look? Ideally I want to merge this soon.

**Comment by ianmcook** [comment]
Thanks @rtpsw & @icexelloss! Is there a test exercising the `"preference": ["POPULATION"]` case? It would be nice to have that.

**Comment by icexelloss** [comment]
> Thanks @rtpsw & @icexelloss! Is there a test exercising the `"preference": ["SAMPLE"]` case? It would be nice to have that.

Sure - are you ok with just copying the json plan (but with preference = SAMPLE) into another test case ?

**Comment by ianmcook** [comment]
> > Thanks @rtpsw & @icexelloss! Is there a test exercising the `"preference": ["SAMPLE"]` case? It would be nice to have that.
> 
> Sure - are you ok with just copying the json plan (but with preference = SAMPLE) into another test case ?

That sounds fine to me!

File: cpp/src/arrow/engine/substrait/extension_set.cc [hunk 2]
```diff
@@ -944,6 +944,30 @@ ExtensionIdRegistry::SubstraitAggregateToArrow DecodeBasicAggregate(
...
+          // https://github.com/substrait-io/substrait/blob/
+          // 73228b4112d79eb1011af0ebb41753ce23ca180c/
+          // extensions/functions_arithmetic.yaml#L1240
+          auto maybe_dist = call.GetOption("distribution");
+          if (maybe_dist) {
+            auto& prefs = **maybe_dist;
+            if (prefs.size() != 1) {
+              return Status::Invalid("expected a single preference for ",
+                                     arrow_function_name, " but got ", prefs.size());
+            }
```
Comment (westonpace) [inline cpp/src/arrow/engine/substrait/extension_set.cc:959]: Technically multiple preferences means the user would prefer the first value but will be ok with the second.  In other words, it's similar to https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Accept

So you should be able to handle this case.  However, no producers do this yet.

**Review by westonpace — APPROVED** [review]
This looks good.  You could maybe handle the multiple-preferences situation, but I don't see that being required for quite a while.  Let me know if you'd rather defer it.

**Pull request merged**